plutus-metatheory/test/TestNEAT.hs

```diff
@@ -3,103 +3,111 @@ module Main where
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN1 <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
+  -- convert Agda normalized type back to named notation:
+  ty1 <- withExceptT FVErrorP $ unDeBruijnTy ty'
+  -- normalize NEAT type using NEAT type normalizer:
+  ty2 <- withExceptT GenError $ convertClosedType tynames k (normalizeTypeG tyG)
 
-  ty2 <- withExceptT GenErrorP $ toClosedType k (normalizeTypeG tyG)
-  return (ty1 == (AlexPn 0 0 0 <$ ty2))
+  -- 4. check the Agda type normalizer agrees with the NEAT type normalizer:
+  unless (ty1 == ty2) $
+    throwCtrex (CtrexNormalizeConvertCommuteTypes k tyG ty1 ty2)
 
-prop_normalizeTypeSame :: TyProp
-prop_normalizeTypeSame k tyG tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN1 <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
 
-  ty2 <- withExceptT TypeErrorP $ unNormalized <$> normalizeType ty
-  return (ty1 == (AlexPn 0 0 0 <$ ty2))
 
-prop_kindInferSame :: TyProp
-prop_kindInferSame k tyG tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
+-- one term-level test to rule them all
+prop_Term :: ClosedTypeG -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_Term tyG tmG = do
+  -- get a production named type
+  ty <- withExceptT GenError $ convertClosedType tynames (Type ()) tyG
+  -- get a production de Bruijn type
   tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  k' <- withExceptT AgdaErrorP $ case inferKindAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just k' -> return k'
-    Nothing -> throwError ()
-  k'' <- withExceptT TypeErrorP $ inferKind defConfig (True <$ ty)
-  return (unconvK (unDeBruijnifyK k') == k'')
+  -- get a production named term
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  -- get a production de Bruijn term
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+
+  -- 1. check the term in the type
+  withExceptT (const $ Ctrex (CtrexTypeCheckFail tyG tmG)) $ liftEither $
+    checkTypeAgda tyDB tmDB
+
+  -- 2. run production CK against metatheory CK
+  tmPlcCK <- withExceptT CkP $ liftEither $ evaluateCk mempty tm
+  tmCK <- withExceptT (const $ Ctrex (CtrexTermEvaluationFail tyG tmG)) $
+    liftEither $ runCKAgda tmDB
+  tmCKN <- withExceptT FVErrorP $ unDeBruijnTerm tmCK
+  unless (tmPlcCK == tmCKN) $
+    throwCtrex (CtrexTermEvaluationMismatch tyG tmG [tmPlcCK,tmCKN])
+
+  -- 3. run all the metatheory evaluators against each other. Taking
+  -- care to normalize the types in the output of runCKAgda. The other
+  -- versions return terms with already normalized types.
+  let evs = [runLAgda,runCKAgda >=> normalizeTypeTermAgda,runTCKAgda,runTCEKVAgda,runTCEKCAgda]
+  let tmEvsM = evs <*> [tmDB]
```

Comment:
Cool kinds write `sequence evs tmDB`, but your version is more readable. I think `map ($ tmDB) evs` would also be readable, but I don't favor any of the options.

Reply:
thanks, `map ($ tmDB) evs` is the cleanest one to me, I'll go with that.